src/meta/meta_ckpt.c

```diff
@@ -918,9 +918,14 @@ __wt_meta_sysinfo_set(WT_SESSION_IMPL *session)
         WT_ERR(__wt_metadata_update(session, WT_SYSTEM_CKPT_URI, buf->data));
     }
 
-    /* We also need to record the oldest timestamp in the metadata so we can set it on startup. */
-    __wt_timestamp_to_hex_string(
-      S2C(session)->txn_global.checkpoint_oldest_timestamp, hex_timestamp);
+    /*
+     * We also need to record the oldest timestamp in the metadata so we can set it on startup. We
+     * should set the checkpoint's oldest timestamp as the minimum of the current oldest timestamp
+     * and the checkpoint timestamp.
+     */
+    __wt_timestamp_to_hex_string(WT_MIN(S2C(session)->txn_global.oldest_timestamp,
```

Comment:
I'm a bit uncomfortable about this. I *think* the shared `txn_global::oldest_timestamp` could be read twice as part of this operation, and if that happened, we might store a value newer than `meta_ckpt_timestamp`, and I think that would be unexpected.
I believe you need a read of oldest into a local variable, followed by a read barrier, then the comparison to be safe here.

Reply:
I'm not sure what you mean. These operations are at the final stage of a checkpoint. Is it possible another checkpoint can start concurrently?

Reply:
Oh, I get it. Yes, you are correct. I don't know it may be read twice.

Reply:
I think I get it. But I think the read_barrier is not necessary.